### PR TITLE
[Triton/Gluon] [Kernel] Add packed FP4 logical transpose

### LINKED ISSUE #5537: Add packed FP4 logical transpose
## Problem

AITER has a fast transpose for ordinary tensor elements, but it does not have a
nibble-aware transpose for matrices stored as two FP4 values per byte. Applying
the existing byte transpose to a packed `(M, N / 2)` tensor produces the wrong
logical shape and nibble pairing.

Training callers currently have to unpack, transpose, and repack with several
framework operations or maintain a duplicate downstream kernel.

## Proposed change

Add a Triton `transpose_packed_fp4` operation that:

- accepts row-major packed FP4 data with physical shape `(M, N / 2)`;
- preserves the low-nibble/even-column and high-nibble/odd-column convention;
- returns the bit-exact logical transpose with shape `(N, M / 2)`;
- supports both raw `uint8` storage and PyTorch's native packed FP4 dtype; and
- handles non-power-of-two dimensions and strided row-major inputs.

## Scope

This issue covers only the row-major packed-FP4 logical transpose, its public
wrapper, correctness tests, and benchmark. It does not add quantization,
dequantization, scale transforms, RHT/Hadamard operations, or a duplicate GEMM
weight shuffle. Callers that need a GEMM-specific physical layout should compose
this operation with AITER's existing weight shuffle.

### PR DESCRIPTION
**Pull request by DaiXindi-AMD**
## Summary

- Add `transpose_packed_fp4` for bit-exact logical transpose of row-major FP4 data packed as two nibbles per byte.
- Keep the kernel, wrapper, tests, and benchmark in the existing `fast_transpose` files.
- Support arbitrary input strides, non-power-of-two tails, `torch.uint8`, and `torch.float4_e2m1fn_x2`.
- Reuse the existing 32 x 32 logical transpose tile and provide a kernel `repr` containing both tile parameters.

This is not a duplicate of `shuffle_weight_f4` or `shuffle_mxfp8fp4_b`: those operations rearrange 16 x 16 packed-byte tiles for GEMM layouts, while this operation produces the logical shape `(N, M / 2)` and repacks nibble pairs after transposition.

Closes #5537.

## Validation

Correctness on MI350X / gfx950 with ROCm 7.2:

```bash
GPU_ARCHS=gfx950 PYTHONPATH=. pytest -q \
  op_tests/triton_tests/quant/test_fast_transpose.py
```

Result: `48 passed`.

The packed-FP4 cases cover nibble order, non-power-of-two tails, strided input, 64-bit stride arguments, native packed dtype preservation, validation, and round trips.

Static checks:

```bash
black --check --fast \
  aiter/ops/triton/_triton_kernels/quant/fast_transpose.py \
  aiter/ops/triton/quant/__init__.py \
  aiter/ops/triton/quant/fast_transpose.py \
  op_tests/triton_tests/quant/test_fast_transpose.py \
  op_tests/op_benchmarks/triton/bench_fast_transpose.py
ruff check \
  aiter/ops/triton/_triton_kernels/quant/fast_transpose.py \
  aiter/ops/triton/quant/__init__.py \
  aiter/ops/triton/quant/fast_transpose.py \
  op_tests/triton_tests/quant/test_fast_transpose.py \
  op_tests/op_benchmarks/triton/bench_fast_transpose.py
python -m compileall -q \
  aiter/ops/triton/_triton_kernels/quant/fast_transpose.py \
  aiter/ops/triton/quant/__init__.py \
  aiter/ops/triton/quant/fast_transpose.py \
  op_tests/triton_tests/quant/test_fast_transpose.py \
  op_tests/op_benchmarks/triton/bench_fast_transpose.py
git diff --check upstream/main...HEAD
```

Result: Black, Ruff, compileall, and `git diff --check` passed.

Benchmark reproducer:

```bash
GPU_ARCHS=gfx950 PYTHONPATH=. \
  python op_tests/op_benchmarks/triton/bench_fast_transpose.py \
  --op packed_fp4 --metric time
```

| Logical shape | PyTorch (ms) | AITER (ms) | Speedup |
|---|---:|---:|---:|
| 6144 x 4096 | 0.349391 | 0.064508 | 5.42x |
| 4096 x 4096 | 0.324700 | 0.057674 | 5.63x |
| 24576 x 4096 | 0.938870 | 0.103893 | 9.04x |
| 4096 x 12288 | 0.600084 | 0.094868 | 6.33x |

The benchmark uses 25 warmup iterations and 100 timed repetitions. A maintainer needs to add `ci:triton-300x` for gfx942 coverage; fork authors cannot add repository labels.

## Checklist

- [x] Kernel has `repr`.
- [x] Unit tests are included.
- [x] Benchmark is included.
- [x] Files follow the existing quant/fast-transpose layout.
- [x] Existing transpose and FP4 shuffle implementations were checked for duplicate functionality.
- [x] No new common helper or third-party dependency.
- [x] No breaking API change.
- [x] Signed off under the project contribution policy.

### CODE REVIEW

File: aiter/ops/triton/_triton_kernels/quant/fast_transpose.py [hunk 1]
```diff
@@ -46,3 +50,62 @@ def _transpose_2d_kernel(
...
+
+    offsets_m = tl.cast(pid_m, tl.int64) * BLOCK_M + tl.arange(0, BLOCK_M)
+    offsets_n_packed = tl.cast(pid_n, tl.int64) * BLOCK_N_PACKED + tl.arange(
+        0, BLOCK_N_PACKED
+    )
+    input_mask = (offsets_m[:, None] < M) & (offsets_n_packed[None, :] < N_PACKED)
+    packed = tl.load(
+        input_ptr
+        + offsets_m[:, None] * stride_input_m
+        + offsets_n_packed[None, :] * stride_input_n,
```
[the remark this answers is not in the document]
Comment (DaiXindi-AMD) [reply aiter/ops/triton/_triton_kernels/quant/fast_transpose.py:89]: Fixed in 7ef865b76: all four input/output strides are explicitly cast to tl.int64 before address arithmetic. Added a GPU regression using singleton dimensions with strides greater than 2**32 for both input and output. Packed transpose tests: 16 passed.

File: aiter/ops/triton/_triton_kernels/quant/fast_transpose.py [hunk 2]
```diff
@@ -46,3 +50,67 @@ def _transpose_2d_kernel(
...
+    low_nibbles, high_nibbles = tl.split(transposed_pairs)
+    output = low_nibbles | (high_nibbles << 4)
+
+    offsets_output_m = tl.cast(pid_n, tl.int64) * BLOCK_N + tl.arange(0, BLOCK_N)
+    offsets_output_n = tl.cast(pid_m, tl.int64) * BLOCK_M_PACKED + tl.arange(
+        0, BLOCK_M_PACKED
+    )
+    output_mask = (offsets_output_m[:, None] < tl.cast(N_PACKED, tl.int64) * 2) & (
+        offsets_output_n[None, :] < tl.cast(M, tl.int64) // 2
+    )
```
Comment (azaidy) [inline aiter/ops/triton/_triton_kernels/quant/fast_transpose.py:109]: Why not use BLOCK_N and BLOCK_M_PACKED here as the mask bounds?

File: aiter/ops/triton/quant/fast_transpose.py [hunk 3]
```diff
@@ -4,13 +4,24 @@
...
-__all__ = ["fast_transpose_2d"]
+__all__ = ["fast_transpose_2d", "transpose_packed_fp4"]
 
 _LOGGER = AiterTritonLogger()
 
+# Both entry points use the same fixed 32x32 logical transpose tile.  For the
+# packed path, the second compile-time extent is half as large because each
+# input byte contains two logical columns.  The shape-dependent ``min`` below
+# only keeps ``tl.arange`` legal for small dimensions; this is not a separate
+# architecture-specific tuning table.
```
Comment (azaidy) [inline aiter/ops/triton/quant/fast_transpose.py]: Can you reduce the comment size? A simple "NOTE: Using fixed 32x32 tile" should be ok

**Review by azaidy — CHANGES_REQUESTED** [review]
Can you move the bench and test into the fast transpose file that already exists. If the kernel is part of the same file then the bench and test should be combined too.

File: aiter/ops/triton/_triton_kernels/quant/fast_transpose.py (hunk 2 above)
Comment (DaiXindi-AMD) [reply aiter/ops/triton/_triton_kernels/quant/fast_transpose.py:109]: These are global offsets, so the mask bounds need to match the packed output shape (2 * N_PACKED, M // 2). Using BLOCK_N and BLOCK_M_PACKED would mask out valid tiles beyond the first tile. Do you mean expressing the mask in tile-local coordinates? That would work, but the bounds would still need to account for the remaining output extent in each tile.

Comment (DaiXindi-AMD) [reply aiter/ops/triton/_triton_kernels/quant/fast_transpose.py]: No code change should be needed here, but I can add a short clarification if you prefer.

**Comment by DaiXindi-AMD** [comment]
Could a maintainer please add the `ci:triton-300x` label? Local runtime validation covered gfx950; this packed-FP4 transpose also needs gfx942 coverage before merge.

File: op_tests/triton_tests/quant/test_fast_transpose.py [hunk 4]
```diff
@@ -46,3 +49,122 @@ def test_fast_transpose_2d_correctness(M, N, dtype):
...
+    output = transpose_packed_fp4(data_fp4)
+
+    torch.testing.assert_close(
+        output, _reference_packed_fp4_transpose(data_fp4), atol=0, rtol=0
+    )
+
+
+def test_transpose_packed_fp4_accepts_int64_strides():
+    huge_stride = (1 << 32) + 1
+    input_storage = torch.tensor([0x21, 0x43], dtype=torch.uint8, device="cuda")
```
Comment (azaidy) [inline op_tests/triton_tests/quant/test_fast_transpose.py:124]: Does this test the input stride? Given packed_N=1 it seems like in the kernel it would be huge_stride*0. Maybe add another uint8 element to the N dim?

**Review by azaidy — APPROVED** [review]
Small note on test case otherwise LGTM

File: op_tests/triton_tests/quant/test_fast_transpose.py (hunk 4 above)
Comment (DaiXindi-AMD) [reply op_tests/triton_tests/quant/test_fast_transpose.py]: Thanks, I’ve addressed that.
